notebook/notebookapp.py

```diff
@@ -1782,7 +1801,11 @@ def cleanup_kernels(self):
         n_kernels = len(self.kernel_manager.list_kernel_ids())
         kernel_msg = trans.ngettext('Shutting down %d kernel', 'Shutting down %d kernels', n_kernels)
         self.log.info(kernel_msg % n_kernels)
-        self.kernel_manager.shutdown_all()
+        # If we're using async kernel management, we need to invoke the async method via the event loop.
+        if isinstance(self.kernel_manager, AsyncMappingKernelManager):
+            asyncio.get_event_loop().run_until_complete(self.kernel_manager.shutdown_all())
+        else:
+            self.kernel_manager.shutdown_all()
```

Comment:
Maybe you want to be more cautious, and use the convenient [run_sync](https://github.com/jupyter/jupyter_server/blob/a2610d33be535b42e37e3c200bcec6afeec341d8/jupyter_server/utils.py#L403-L442) we will use in jupyter_server:
```python
run_sync(self.kernel_manager.shutdown_all())
```